### PR TITLE
[WP Stories] fix stories editor session start

### LINKED ISSUE #15291: Adjust postEditorAnalyticsSession calls to make startup_time_ms meaningful for Stories Editor
### Expected behavior

For different editors in WPAndroid we us a `postEditorAnalyticsSession` object and `wpandroid_editor_session_start`  tracks event to track editor startup times.  

For the gutenberg editor we have described the `startup_time_ms` property for gutenberg as: "[`startup_time_ms`] is the representation of the loading time (on Android, time measurement starts during the activity onCreate and ends when we render Gutenberg content)."

For reference, the `wpandroid_editor_session_start` event is sending the following properties. 
<details>
<summary>Event Properties</summary>
<br>
blog_type	

has_unsupported_blocks	

device_info_model	

device_info_os	

device_info_os_version	

device_info_phone_radio_type	

device_info_wifi_connected	

editor	

editor_has_hw_disabled	

post_type	

device_info_language	

session_id	

startup_time_ms	

template	

unsupported_blocks	

user_info_app_scheme	

user_info_dotcom_user	

user_info_gutenberg_enabled	

user_info_jetpack_user	

user_info_number_of_blogs	

device_info_manufacturer	

device_info_is_rtl_language	

browser_type	

device_info_current_network_operator	

can_view_editor_onboarding	

content_type	

device_info_app_name	

device_info_app_version	

device_info_app_version_code	

device_info_bluetooth_enabled	

device_info_bluetooth_version	

device_info_brand	

device_info_device_orientation	

device_info_is_online	

device_info_display_density_dpi	

device_info_display_height	

device_info_display_size	

device_info_display_usable_height	

device_info_display_usable_width	

device_info_display_width	

device_info_has_nfc	

device_info_has_telephony	

device_info_is_large_display

</details>

### Actual behavior

Currently there is a problem with the `startup_time_ms` property when sent with the `wpandroid_editor_session_start` event during Stories Editor startup. The call to create the session and the code for calculting startup time are essentially right after each other making the measurement not useful: 

https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerViewModel.kt#L113-L117

For the stories editor, we should likewise not count the session as started with the `startup_time_ms` calculation until the stories editor is as close to showing as possible. In android terminology we probably dont want this counting as started until onResume lifecycle event is called, but more importantly we want the measurement to be meaningful and not always be zero. 

### Steps to reproduce the behavior
*recommend watching tracks live view for `wpandroid_editor_session_start` event for your test username if you have access to tracks ([example url](https://mc.a8c.com/tracks/live/?eventname=wpandroid_editor_session_start&user=cameron.voell%40automattic.com&useragent=)). Otherwise you can just debug with a breakpoint at the line [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerViewModel.kt#L117) and note that the `startup_time_ms` calculation is not meaningful. 

1. While logged on as a wpcom user use the add button and select Story Post
2. Choose any Media and click check mark
3. Verify the zero `startup_time_ms` via tracks or debugger

##### Tested on [device], Android [version], WPAndroid [version]
Pixel 4 Emulator API 30

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #15291 

See accompanying stories library PR https://github.com/Automattic/stories-android/pull/714

**tl;dr** the session was being marked as started right after having created the `PostEditorAnalyticsSession` object, which invariably meant an elapsed startup time of 0 to 1 milliseconds when looking into Tracks.

This PR aims at fixing this by adding a few methods and making the calls at the right moments, which are after having loaded a Story and the composer is about to show it on the screen.

### Description

Following up on this [comment](https://github.com/wordpress-mobile/WordPress-Android/issues/15291#issuecomment-915257034), the moments in which Story frames are loaded onto the Story composer can be roughly grouped as follows:

#### MediaPicker
Using the **MediaPicker**. The `EditorMediaListener`'s `appendMediaFiles` is the last entry point to adding Story slides once the media has been obtained from the media picker and passed on to the Story Composer.  In the StoryComposerActivity, we have an observer on `mediaFilesUris` object which is ultimately reponsible for adding the media items as slides, via the `addFramesToStoryFromMediaUriList()` method:

```
    private fun setupViewModelObservers() {
        viewModel.mediaFilesUris.observe(this, { uriList ->
            val filteredList = uriList.filterNot { MediaUtils.isGif(it.toString()) }
            if (filteredList.isNotEmpty()) {
                addFramesToStoryFromMediaUriList(filteredList)
                setDefaultSelectionAndUpdateBackgroundSurfaceUI(filteredList)
            }
```

So, adding a call at that point to signal the session start (or what is the same, readiness of the editor for user interaction) is the right way to measure it (added [here](https://github.com/wordpress-mobile/WordPress-Android/compare/issue/15291-stories-editor-session-start?expand=1#diff-32ef15fee2cf2c96f64279e6c70d35a31552e6508585386c60e57ef2dd19b21dR229-R233)).
This part is executed for any media items chosen with the media picker (local to the device,  from other apps, from the free photo library, or from the WordPress site's media).

_This part covers most of the use cases that are of interest or are comparable to the time-to-start being measured for Aztec or Gutenberg._

#### Starting the composer on live camera preview mode
If the user chooses to tap on the camera FAB presented on the picker, which triggers the live preview mode of the Story Composer, then things get a bit trickier. In reality, the moment when things are ready to be presented in this mode is detected either in [`BackgroundSurfaceManagerReadyListener`](https://github.com/Automattic/stories-android/blob/develop/stories/src/main/java/com/wordpress/stories/compose/ComposeLoopFrameActivity.kt#L504-L518) or within `launchCameraPreviewWithSurfaceSafeguard`, both internal to ComposeLoopFrameActivity in the Stories library.

I'd argue this second moment is essentially not as important, given this is very specific to the Story composer, so in this case I'd stick to something simpler at the client app level (StoryComposerActivity), and just check whether we've been passed the `requestCodes.EXTRA_LAUNCH_WPSTORIES_CAMERA_REQUESTED  in `handleMediaPickerIntentData`. This is done [here](https://github.com/wordpress-mobile/WordPress-Android/compare/issue/15291-stories-editor-session-start?expand=1#diff-32ef15fee2cf2c96f64279e6c70d35a31552e6508585386c60e57ef2dd19b21dR389-R391).

#### Loading an existing Story
When the user loads an existing Story (by tapping on a Story block in GB mobile that was already created on this very device), then we have a slightly different situation, in which the actual loading is taken care of via Intent and handled internally by ComposeLoopFrameActivity's `onLoadFromIntent()`. For this then, we added `StoryLoadStart` and `StoryLoadEnd` events in the stories library in https://github.com/Automattic/stories-android/pull/714; that can be listened to from the client app in order to do whatever is needed (in this case, to mark the start of the user editing session once we get a `StoryLoadEnd` event - see https://github.com/wordpress-mobile/WordPress-Android/commit/2132127d12d3aa43f5e93bd7e61a99a19ec3424a).

#### Special case: editing an empty Story block
If the user adds an empty Story block to a Gutenberg post and then taps on ADD MEDIA, or edits an existing block ~things work the same here.~ things are a bit different: 
- The StoryComposerActivity gets launched, and it realizes it needs to show the provided Media Picker so to allow the user to add media
- so now the MediaPicker is shown, waiting for the user's pick results
- finally, we get a call in `onActivityResult`, and only at this point and after adding the media slides (see the section above called MediaPicker, which covered how this part of the flow works) we know how much time it took for the StoryComposer to set itself up with the media it was passed along from the Media Picker.

Given this, we added a new method to PostEditorAnalyticsSession in order to [reset the `mStartTime`](https://github.com/wordpress-mobile/WordPress-Android/compare/issue/15291-stories-editor-session-start?expand=1#diff-471b17a541b67be8b228ee74e749614c81c28bed72dad82663e442b19d876c5eR131-R138) to current time when we come back from the media picker, given we really want to measure the StoryComposer start up time since getting the media items until after these have been added as media slides (otherwise, we'd be counting also all the time spent by the user choosing media items in the MediaPicker).

#### Final note
It's worth noting that the session's object `start()` method may be called more than once when you keep adding slides to a Story at different moments (by tapping the "+" button on the StoryFrameSelector at the bottom-right of the screen). Fortunately there's a [safeguard within the method itself,](https://github.com/wordpress-mobile/WordPress-Android/blob/daecb9589521445025ee3b66e5636cd1ecee1aad/WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java#L125-L128) that doesn't allow that to happen and gets an entry on the application's log with a warning. This is fine and I'm relying on this behavior to avoid having more complicated checks at the client app level that would make things only more complex.

### To test:

**TEST CASE A: Select media from media picker**
1. create a new story
2. when the media picker appears, select one or more media items
3. Observe the media items get added, and the logs show an entry such as this:
```
2021-09-08 19:50:10.787 29365-29365/org.wordpress.android I/WordPress-STATS: 🔵 Tracked: editor_session_start, Properties: {"has_unsupported_blocks":"0","editor":"wp_stories_creator","content_type":"new","startup_time_ms":1054,"session_id":"e9df54a2-1e1b-4fa5-8078-04edf0892f46","post_type":"post","blog_type":"wpcom","editor_has_hw_disabled":"0","unsupported_blocks":[]}
```

4. tap on the "+" button on the story composer's story frame selector to add more media items
5. pick one or more media items
6. observe the new item(s) get added as slides, and the following entry is observed in the log:

```
2021-09-08 19:51:22.062 29365-29365/org.wordpress.android W/WordPress-EDITOR: An editor session cannot be attempted to be started more than once, unless it's due to rotation or Editor switch
```

(that's expected)

7. once more, tap on the "+" button on the story composer's story frame selector to add more media items
8. when the media picker appears, cancel it (tap on the back arrow)
9. observe the following log entry:

```
2021-09-08 19:52:23.574 29365-29365/org.wordpress.android W/WordPress-EDITOR: An editor session start time cannot be reset once it's started
```

That's expected, since the session has already been started the first time we added media items.

**TEST CASE B:** 
- repeat case A, but in step 2 try choosing media from the local device, free photo library, and the site's WP Media.

Picking pictures that are not local to the device (for example, from Google Photos or other apps) may take a while, but this time is not considered as part of the StoryComposer load time. So, when adding media items like this, the logs should show something like:

```
2021-09-08 19:58:31.935 29365-29365/org.wordpress.android I/WordPress-STATS: 🔵 Tracked: editor_session_start, Properties: {"has_unsupported_blocks":"0","editor":"wp_stories_creator","content_type":"new","startup_time_ms":647,"session_id":"7230c859-4c9a-4745-a52e-a50f25c5ff1f","post_type":"post","blog_type":"wpcom","editor_has_hw_disabled":"0","unsupported_blocks":[]}
```

**TEST CASE C: existing empty block**
1. create a new post in GB mobile, and add a Story block
2. tap on ADD MEDIA on the Story block.
3. when the media picker appears, select one or more media items to add
4. observe these get added, and a log entry such as this appears:

```
2021-09-08 20:01:49.164 29365-29365/org.wordpress.android I/WordPress-STATS: 🔵 Tracked: editor_session_start, Properties: {"has_unsupported_blocks":"0","editor":"wp_stories_creator","content_type":"new","startup_time_ms":503,"session_id":"c08cd534-c9a2-4a53-910a-2e0e80ad9cf0","post_type":"post","blog_type":"wpcom","editor_has_hw_disabled":"0","unsupported_blocks":[]}
```
5. Save the Story and verify the Story Block gets populated in GB.

**TEST CASE D:** editing existing non-empty Story block, that was created on this device
1. create a block post with a story block with some slides in it
2. open the post and tap on the Story block twice to edit the Story block
3. observe the StoryComposer fires up and shows your slides, and a line appears in the log as well.

## Regression Notes
1. Potential unintended areas of impact
None, this is only related to how we track editor sessions startup timing, and is only restricted to the StoryComposer

2. What I did to test those areas of impact (or what existing automated tests I relied on)
See manual testing paths above

3. What automated tests I added (or what prevented me from doing so)
The nature of startup time is dependant on various entry points which are tightly coupled with the Media Picker and other sources to load Story slides from. The easiest way to test each path is described above.

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by jd-alexander — COMMENTED** [review]
## Entry Points Testing
I enabled `Collect Information` within the Privacy Settings so that the WordPress Stats would be printed to the log output. 

### 1. Select media from media picker

#### Test Case A
- I followed the actions of creating a new story and I observed the respective log output with a realistic `startup_time_ms`.
- I noted the `An editor session cannot be attempted to be started more than once, unless it's due to rotation or Editor switch` log entry when I added more media items to the story.
- I noted the `An editor session cannot be attempted to be started more than once, unless it's due to rotation or Editor switch` log entry when I opened the media picker and exited.

#### Test Case B

- I utilized the local device, free photo library, and WP Media as media sources for the Story Composer. All cases created log output with a realistic `startup_time_ms`.

### 2. Existing Empty Block

#### Test Case C
- I observed the log entry after creating a story block with media within a blog post.

### 3. Loading Existing Story
#### Test Case D

-  I observed the log entry after editing an existing story block with media within a blog post.


### 4. Camera Preview
- I tested the flow from the Camera Preview. I observed that the log output was created when the Camera Preview opened and not when the user returned to the Story Composer. Is this expected?

## Final comments 

Thanks for these changes. Apart from the Camera Preview comment above, this works as expected. Once we clear that up we are good to go! 

#### Special mentions
- I appreciate the usage of EventBus events to propogate story loading start and end events from the underlying ComposeLoopFrameActivity.

> It's worth noting that the session's object `start()` method may be called more than once when you keep adding slides to a Story at different moments (by tapping the "+" button on the StoryFrameSelector at the bottom-right of the screen). Fortunately there's a [safeguard within the method itself,](https://github.com/wordpress-mobile/WordPress-Android/blob/daecb9589521445025ee3b66e5636cd1ecee1aad/WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java#L125-L128) that doesn't allow that to happen and gets an entry on the application's log with a warning. This is fine and I'm relying on this behavior to avoid having more complicated checks at the client app level that would make things only more complex.

- Ah good observation. This was an issue I was concerned about due to the nature of the media picking experience and the related state changes that could trigger the session start/restart.

**Comment by mzorz** [comment]
Thank you so much @jd-alexander for your thorough testing as usual! 🙏  🙇 

> I tested the flow from the Camera Preview. I observed that the log output was created when the Camera Preview opened and not when the user returned to the Story Composer. Is this expected?

Great observation - this is expected indeed, the same way we wouldn't want to reset the editor startup time each time the user adds a new image using the camera in Gutenberg, we don't reset the time here. I take the camera preview mode to be a special case of initialization for the Stories Composer but, essentially, once the camera preview is shown once then we can say the composer has started. For subsequent switches to the camera preview mode (for example, to add more slides) we don't measure startup times any further, because technically it's already loaded 👍 

I updated the stories library hash after adding the comment here https://github.com/Automattic/stories-android/commit/34817a3c1a315e3f139622d2612266e6c52cf6c1, I'll update it here once more after merging that PR, and then this one should then be ready 👍

**Review by jd-alexander — APPROVED** [review]
> Great observation - this is expected indeed, the same way we wouldn't want to reset the editor startup time each time the user adds a new image using the camera in Gutenberg, we don't reset the time here. I take the camera preview mode to be a special case of initialization for the Stories Composer but, essentially, once the camera preview is shown once then we can say the composer has started. For subsequent switches to the camera preview mode (for example, to add more slides) we don't measure startup times any further, because technically it's already loaded 👍

Thanks for expounding! All clear now! LGTM :shipit:
